### PR TITLE
Fix performance regression in macOS wxDataViewCtrl

### PR DESCRIPTION
**Pull request by vslavik**
Fixes *severe* degradation of performance caused by 36ea7ff4d61e174a2d116819f98e66f22a7f5b8a, which forces calculation of fitting size even for non-autosized columns.

The purpose of the change, as per @softwarefailure's explanation under the above commit, was to update row heights as well, but doing that doesn't make sense when the row height is fixed. This PR simply avoids calling `FitColumnWidthToContent()` unless it actually is needed, i.e. the column is autosized or variable row heights are used.

It possibly (probably?) conflicts with PR #2352 by @dkulp but if it is a correct thing to do, it is both simpler *and* more functional fix for the performance regression — i.e. I think they should be both applied.

In [Poedit](https://github.com/vslavik/poedit), a cell in a DVC control is updated every time the user types; it's a single cell, with fixed-sized columns, with ~1000 rows in my tests (files an order of magnitude larger are common; I didn't dare try them). The performance impact was approximately as follows:

1. before 36ea7ff4d61e174a2d116819f98e66f22a7f5b8a: typing in realtime, CPU use under 10% (unrelated processing)
2. 36ea7ff4d61e174a2d116819f98e66f22a7f5b8a: typing with extreme latency, one second between keystrokes, 100% CPU usage
3. with PR #2352: much better, but still  high latency and 50-70% CPU usage
4. with this PR: back to 1., no noticeable latency

### CODE REVIEW

**Comment by vadz** [comment]
Thanks, I do think it's the right thing to do, but I'm very lost in all the changes/discussions related to this, and I never had time to actually test this myself, so I hesitated to change anything in the PRs done by the other people who did test it. But FWIW I wholeheartedly agree with you and I was wondering about this (i.e. why should this be done for non-autosized columns) myself too.

This being said, I think skipping all this when the control is frozen, which is what #2352 does, AFAIK, still makes sense too, so ideally we'd apply both of them. Why exactly does this one conflict with it? Or do you mean it's just a textual, i.e. Git, conflict, which could, hopefully, be just resolved?

In any case, if you've tested this and confirmed that it works in the dataview sample in the different previously discussed scenarios, please do merge it, TIA!

**Review by vadz — APPROVED** [review]

**Comment by vslavik** [comment]
Sorry, I meant it more like "I'm not 100% sure this isn't in conflict" — I'm myself mostly confused by the changes.

Re-reading some parts again, I suspect the calculation is meant for fixed-height case (i.e. to automatically determine the row height).  *If* that's the case, this PR is wrong (but then I don't understand how, if at all, `SetRowHeight()` works).

In any case, yes, #2352 has important optimizations regardless.

**Comment by softwarefailure** [comment]
I don't think this patch should be applied because `wxDV_VARIABLE_LINE_HEIGHT` is a completely different thing. AFAIU `wxDV_VARIABLE_LINE_HEIGHT` allows lines to have individual heights, i.e. each line in a wxDVC may have its individual height. That's not what my patch was about at all. The reason why I was calling `FitColumnWidthToContent()` even for non-autosize columns was to have autosizing also for line height, not just for column width. 

Imagine the following situation: The user unfolds a tree node and the newly visible rows contain some icons that are bigger than the current line height. On MSW and GTK, the line height is now automatically resized so that it can fully show that icon. On macOS, however, the icon will be cut off because the minimum line height is not automatically adapted to the newly unfolded row's height . That's what my patch was about. So if you apply this PR, autosizing of line heights will no longer work on macOS and wxDVC's behaviour on macOS will be inconsistent to MSW and GTK where autosizing of line heights works nicely.

As said above, this autosizing of line heights is **not** the same as `wxDV_VARIABLE_LINE_HEIGHT`. Autosizing of line heights as done by wxDVC on MSW and GTK will not allow individual row heights as `wxDV_VARIABLE_LINE_HEIGHT` does. Instead, it will just make sure that the global row height is automatically increased in case the row's content doesn't fit into the current row height. 

> The purpose of the change, as per @softwarefailure's explanation under the above commit, was to update row heights as well, but doing that doesn't make sense when the row height is fixed. 

It does. See my explanation above.

TLDR; I think #2352 should be applied instead. It fixes the performance issue without breaking autosizing of line heights. #2370 will break autosizing of line heights which is why it shouldn't be applied.

**Comment by dkulp** [comment]
It should also be mentioned that wxDV_VARIABLE_LINE_HEIGHT isn't even implemented on OSX at all.   There is code there to calculate the variable heights, but wxCocoaDataViewControl::SetRowHeight is a no-op that states it's not supported by the native control.

**Comment by softwarefailure** [comment]
> That's a weird corner case (or just plain wrong from UI/UX perspective).  OTOH it degrades perfectly normal and common use, with constant-height rows, sometimes to the point of being unusable.

I'm all for consistency. Even though it might be uncommon, it works correctly on MSW and GTK so we should find a way to make it work on macOS too. Even if it's a little slower. macOS is the epitome of sluggish anyway ;) Users who want the ultimate performance should `Thaw` and `Freeze` then. Could be mentioned in the docs.

> Why is the height being calculated even after an explicit call to `SetRowHeight()`?

AFAIU `SetRowHeight` is just about the *minimum* row height. This doesn't mean that the row height can't grow. At least that's the behaviour on MSW and GTK.

**Comment by vslavik** [comment]
> I'm all for consistency. Even though it might be uncommon, it works correctly on MSW and GTK so we should find a way to make it work on macOS too.

Yes, we should find a way to have consistency, but I'd humbly suggest that a patch that massively degrades performance in the 99.99% of mainstream situations for the benefit of a 0.01% corner case is *not* that way.

> Even if it's a little slower. macOS is the epitome of sluggish anyway ;)

It's not "a little slower". It made common use of wxDVC *orders of magnitude* slower - 12 minutes vs 15 seconds is 48x slower. "A little slower" may be a tradeoff worth doing, 48x slower is a problem and not something to be this blasé about.  In this light, snide remarks about "epitome of sluggish" are perhaps a bit misplaced.

> Users who want the ultimate performance should Thaw and Freeze then.

Freeze/Thaw is not a silver bullet. It helps with batch updates: what used to be O(n) remains O(n) and doesn't go quadratic (that's Daniel's use case).

But your changes degraded performance even of single-row, all-fixed-sizes updates — something that used to be O(1), and is reasonably expected to be, is now O(n) in the size of the model. Adding Freeze/Thaw on top of that would just increase the constant.

#2352 largely addresses this by eliminating the extra O(n) factor of pointless calculations.

**Comment by softwarefailure** [comment]
> It's not "a little slower". It made common use of wxDVC orders of magnitude slower - 12 minutes vs 15 seconds is 48x slower.

Of course I was referring to the patch by @dkulp here, not to my original one. As I said before, I'm all for applying #2352 because it makes things much faster without removing any functionality that I added in my patches. The only thing that should be fixed before applying is the size last column to fit thing.

**Comment by vadz** [comment]
Sorry, I didn't have time to follow everything, so I'm still not sure if we have a consensus now? I.e. can anybody please tell me in simple words what should I do (or, alternatively, Vaclav, could you please explain what are you going to do if you plan on doing it yourself?)? TIA!

**Comment by vslavik** [comment]
This PR should *not* be merged.

It is my understanding-slash-hope that we're waiting on @softwarefailure to submit a patch that fixes the performance regression related to `sizeLastColumnToFit:` as they seem to understand both what's wrong with #2352 and the issues 7555d1b2459949c128d6ee3a473669a1a134f124 and 36ea7ff4d61e174a2d116819f98e66f22a7f5b8a were meant to address.

Ideally, we would then:

1. Merge #2352 + @softwarefailure's patch + my patch from https://github.com/vslavik/wxWidgets/commit/f2324d53ecad8dc88240376225ea1d0969e3384c 

Alternatively:

1. Merge #2352 
2. Apply https://github.com/vslavik/wxWidgets/commit/f2324d53ecad8dc88240376225ea1d0969e3384c
3. Revert 7555d1b2459949c128d6ee3a473669a1a134f124 (logically, `git revert` won't apply cleanly)

Either way I'm happy to prep a clean single PR, but let's give it a few days first.

**Comment by softwarefailure** [comment]
I can try to get rid of sizeLastColumnToFit but I'm currently quite busy with other things so you'd have to wait for some days until I have time to look into it.
